File: Python/366__Find_Leaves_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []
        visited = set()
        while root and not root in visited:
            leaves = []
            self.dfs(root, leaves, visited)
            result.append(leaves)
        return result

    def dfs(self, root, leaves, visited):
        if not root:
            return
        if root.left in visited:
            root.left = None
        if root.right in visited:
            root.right = None
        if not root.left and not root.right:
            leaves.append(root.val)
            visited.add(root)
            return
        self.dfs(root.left, leaves, visited)
        self.dfs(root.right, leaves, visited)

# Faster Solution

class Solution(object):
    def findLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        dic = {}
        self.run(root, dic)
        result = []
        for height in dic:
            result += [dic[height]]
        return result

    def run(self, node, dic):
        if not node:
            return 0

        left = self.run(node.left, dic)
        right = self.run(node.right, dic)

        height = max(left, right)

        dic[height] = dic[height] + [node.val] if height in dic else [node.val]

        return height + 1

File: Java/366__Find_Leaves_of_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> list = new ArrayList<>();
        Map<Integer, List<Integer>> map = new HashMap<>();
        int depth = dfs(root, map);
        for (int i = 0; i <= depth; i++)
            list.add(map.get(i));
        return list;
    }

    private int dfs(TreeNode node, Map<Integer, List<Integer>> result) {
        if (node == null) return -1;
        if (node.left == null && node.right == null) {
            if (!result.containsKey(0)) {
                List<Integer> list = new ArrayList<>();
                list.add(node.val);
                result.put(0, list);
            } else
                result.get(0).add(node.val);
            return 0;
        }
        int depth = Math.max(dfs(node.left, result), dfs(node.right, result)) + 1;
        if (!result.containsKey(depth)) {
            List<Integer> list = new ArrayList<>();
            list.add(node.val);
            result.put(depth, list);
        } else
            result.get(depth).add(node.val);
        return depth;
    }

}
